### PR TITLE
[Core] An access_next method in ShuffleCombiner

### PR DESCRIPTION
**Pull request by xwkuang5**
issue #49
Implement shuffle combiner with ZeroMQ publisher-subscriber pattern.
Each worker will be both publisher and subscriber at the same time.

### CODE REVIEW

**Comment by ddmbr** [comment]
How does the performance look like? Can you try the PageRank benchmark (in `benchmarks/`) with the twitter dataset (on `/datasets/graph/twitter` in the HDFS cluster you are using) and share with me the numbers? Thanks!

**Comment by kygx-legend** [comment]
And about the PageRank benchmark, could you help modify it with the inputformat factory to create inputformat?

File: core/shuffle_combiner.hpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+#pragma once
+
+#include <atomic>
+#include <string>
+#include <vector>
+
+#include "base/serialization.hpp"
+#include "base/thread_support.hpp"
+#include "zmq_helpers.hpp"
```
Comment (kygx-legend) [inline core/shuffle_combiner.hpp]: `core/zmq_helpers.hpp`

File: core/shuffle_combiner.hpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        local_id_ = sc.local_id_;
+        sc.num_units_ = 0;
+        sc.num_local_workers_ = 0;
+        sc.channel_id_ = -1;
+        sc.local_id_ = -1;
+        sc.publisher_ = nullptr;
+        sc.subscriber_ = nullptr;
+    }
+
+    ~ShuffleCombiner() {
```
Comment (kygx-legend) [inline core/shuffle_combiner.hpp]: `virtual`

File: core/shuffle_combiner_factory.hpp [hunk 3]
```diff
@@ -19,7 +19,8 @@
 #include <utility>
 #include <vector>
 
-#include "core/shuffler.hpp"
+#include "shuffle_combiner.hpp"
```
Comment (kygx-legend) [inline core/shuffle_combiner_factory.hpp]: missing `core` directory

File: core/shuffle_combiner_unittest.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
+#include "core/shuffle_combiner.hpp"
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "zmq_helpers.hpp"
```
Comment (kygx-legend) [inline core/shuffle_combiner_unittest.cpp]: #include "core/zmq_helpers.hpp"

File: core/shuffle_combiner_unittest.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace husky {
+namespace core {
+
```
Comment (kygx-legend) [inline core/shuffle_combiner_unittest.cpp]: missing anonymous namespace

File: core/shuffle_combiner_unittest.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+#include <chrono>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "zmq_helpers.hpp"
+
+namespace husky {
+namespace core {
```
Comment (kygx-legend) [inline core/shuffle_combiner_unittest.cpp]: no namespace named core.

File: core/shuffle_combiner_unittest.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    TestShuffleCombiner() {}
+    ~TestShuffleCombiner() {}
+
+   protected:
+    void SetUp() {}
+    void TearDown() {}
+};
+
+TEST_F(TestShuffleCombiner, ShuffleCombiner) {
+    const int NUM_GLOBAL_WORKERS = 4, NUM_LOCAL_WORKERS = 2, ROUND = 3, MAX_TIME = 10;
```
Comment (kygx-legend) [inline core/shuffle_combiner_unittest.cpp]: For local variables, don't use uppercases.

File: core/shuffle_combiner.hpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    int access_next() {
+        int worker_id = husky::zmq_recv_int32(subscriber_);
+        return worker_id;
+    }
+
+    CollectT& storage(unsigned idx) { return messages_[idx].storage(); }
+
+    static void init_sockets(int num_local_workers, int channel_id, zmq::context_t& context) {
+        if (channel_sc_socket_map_.find(channel_id) == channel_sc_socket_map_.end()) {
```
Comment (zzxx-husky) [inline core/shuffle_combiner.hpp]: Maybe you change to share the two sockets among different channels in the future.

Comment (xwkuang5) [inline core/shuffle_combiner.hpp]: Okay, maybe we can send two integers one time? the first one is used to identify the channel that should be receiving message and the second integer  is used to specify which worker is available.

**Comment by ddmbr** [comment]
How's the status of this one?

**Comment by xwkuang5** [comment]
Do you mean the version that all channels share the same two sockets? I haven't implemented that one yet. The version in this PR is the same as the one that I told you about in the issue page.

**Comment by ddmbr** [comment]
@xwkuang5 You need to rebase

**Comment by xwkuang5** [comment]
Please review @ddmbr @kygx-legend

**Pull request merged**